Updated Microsoft.CodeAnalysis.PublicApiAnalyzers to latest stable

BrennanConroy opened the pull request.

dougbu commented on src/Middleware/CORS/src/PublicAPI.Shipped.txt:
```diff
@@ -2,8 +2,6 @@
 ~Microsoft.AspNetCore.Cors.CorsPolicyMetadata.CorsPolicyMetadata(Microsoft.AspNetCore.Cors.Infrastructure.CorsPolicy policy) -> void
 ~Microsoft.AspNetCore.Cors.CorsPolicyMetadata.Policy.get -> Microsoft.AspNetCore.Cors.Infrastructure.CorsPolicy
 ~Microsoft.AspNetCore.Cors.EnableCorsAttribute.EnableCorsAttribute(string policyName) -> void
-~Microsoft.AspNetCore.Cors.EnableCorsAttribute.PolicyName.get -> string
-~Microsoft.AspNetCore.Cors.EnableCorsAttribute.PolicyName.set -> void
```
Odd, looks like we somehow had a fair number of true duplicate lines in this file. How did we get away w/ that❔

BrennanConroy replied:
Because the analyzer sucks.

dougbu reviewed (commented):
I'm not sure how to review this thoroughly since the changes mix the different cases. Might be worthwhile splitting out (for example) the removal of true duplicate lines from the rest. The next stage might be to handle nullability changes for existing methods (remove a large number of `*REMOVED*` lines from the Unshipped files while annotating the Shipped files correctly). Finally, could do the Microsoft.CodeAnalysis.PublicApiAnalyzers version bump while adding default constructors.

Another debatable issue is how to handle default constructors we shipped in .NET 6.0.0. They all show up in PublicAPI.Unshipped.txt files in this PR.

dougbu commented:
This is actually very close to passing CI validation. Congrats @BrennanConroy❕

Should decide a few things like
1. whether we instead want fixes for the PublicAPI analyzer regressions instead (see #41100)
2. where to place default constructors we shipped in .NET 6.0.0, and
3. how to get a thorough review

DamianEdwards commented:
Any update on if when we might get this merged?

BrennanConroy commented:
> 1. whether we instead want fixes for the PublicAPI analyzer regressions instead

The only real regression IMO is the forwarded types being messy. I think the issue with *REMOVED* for non-nullable annotated APIs when we annotate them as nullable is fine, it should just be this one time cost.

> 2\. where to place default constructors we shipped in .NET 6.0.0

I don't feel strongly about figuring out the "correct" solution to this, I think it's completely fine to leave them in the unshipped file and move on with our lives.

> 3\. how to get a thorough review

Good luck 😄 , I'm not going to spend a couple hours trying to have separate commits for "removing duplicates", adding default ctors, etc.

BrennanConroy committed:
Adding default ctors

BrennanConroy committed:
Remove duplicates

BrennanConroy committed:
normalize nullable

BrennanConroy commented:
PR updated with split commits.
1. Added default ctors to unshipped files
2. Removed duplicate entries (Same APIs but one with ~ and one without ~)
3. Normalized nullability in APIs

The third one is a bit odd since there are still APIs with ~, basically just kept modifying based on the analyzers errors over and over until the build passed. I think we should remove all ~ from our shipped files at the end of 7.0.

dougbu commented:
Thanks for the splitting @BrennanConroy❕ I'll try to look at this today but would also like to hear from @DamianEdwards, @dotnet/aspdoi and / or @dotnet/aspnet-build on the other questions.

@jmarolf any chance the package will be or has been fixed enough to avoid these shenanigans❔

jmarolf commented:
> @jmarolf any chance the package will be or has been fixed enough to avoid these shenanigans❔

Which shenanigans? We accept PRs :). icrosoft.CodeAnalysis.PublicApiAnalyzers is just infrastructure that roslyn wrote to meet our own needs.

dougbu commented:
> > @jmarolf any chance the package will be or has been fixed enough to avoid these shenanigans❔

> Which shenanigans? We accept PRs :) Microsoft.CodeAnalysis.PublicApiAnalyzers is just infrastructure that roslyn wrote to meet our own needs.

That's not great for a public project we and others have taken a hard dependency on.

The tooling behaves rather differently in v3.3.0 and v.3.3.3. We didn't make those changes.

@dotnet/aspnet-build @dotnet/aspdoi any suggestions for alternative public API verifiers❔ We can't go back to `ApiCompat` because that product has too many bugs and restrictions. And PublicApiAnalyzers v3.3.0 can't handle #41100. I'm thinking this PR is a lot of work without a future.

😦

jmarolf commented:
Probably also worth discussing why the supported features for this don't meet your needs https://docs.microsoft.com/dotnet/fundamentals/package-validation/overview

wtgodbe commented:
Doesn't package validation only confirm that a package has the same surface area across the TFMs & RIDs it supports? We also need to verify that we don't change API surface in servicing branches, which I'm not sure the package validation tool does.

wtgodbe commented:
> Added default ctors to unshipped files

Do you mean added to .shipped files (for ones from pre-7.0)? I'd be alright with this PR if that was the case

dougbu commented:
> Probably also worth discussing why the supported features for this don't meet your needs https://docs.microsoft.com/dotnet/fundamentals/package-validation/overview

We have that enabled in this repo, have a few CompatibilitySuppressions.xml files, and enable baseline version validation in servicing builds but&hellip;

1. The [baseline-version-validator](https://docs.microsoft.com/en-ca/dotnet/fundamentals/package-validation/baseline-version-validator) doesn't track incremental API additions
2. The [baseline-version-validator](https://docs.microsoft.com/en-ca/dotnet/fundamentals/package-validation/baseline-version-validator) can't prevent public API additions in servicing (it focuses on breaking changes, not all changes)
3. Our CompatibilitySuppressions.xml would be _huge_ if they listed all of the nullability changes we've made since the 6.0.0 API shipped

dougbu commented:
@jmarolf for some history here, this goes back past #24188. @Pilchie let @mavasani know our plans to use Microsoft.CodeAnalysis.PublicApiAnalyzers before we started work on that PR. @sharwell was also involved.

dougbu reviewed (commented):
Am I correct all of the "duplicate" entries removed in the second commit are shipped signatures lacking all nullable annotations, `*REMOVED*` signatures for the same, or `*REMOVED*` signatures containing a leading `~` but also the correct nullable annotations❔ If that's the case, the analyzer must previously have ignored the nullable annotations when eliminating duplicates. I can see that as fixing a bug but the default constructor thing isn't great and I haven't looked at the third commit.

wtgodbe commented:
@SteveSandersonMS @mattleibow has dotnet/maui had any issues with the 3.3.3 version of `Microsoft.CodeAnalysis.PublicApiAnalyzers`?

dougbu commented:
@SteveSandersonMS @mattleibow see also https://github.com/dotnet/aspnetcore/pull/41115#issuecomment-1108921992

dougbu commented:
@BrennanConroy did you test whether changing another file to use nullable annotations and using the analyzer's code fix introduces new duplicates in the PublicApi.Shipped.txt file❔ I'm concerned not only about the next patch that contains a fix we need and further big breaks but also about having to fight the analyzer if we move to v3.3.3.

BrennanConroy commented:
Nope. Since you obviously have strong opinions about how the analyzer should behave, I suggest you try it out. I don't really care as long as it isn't a really bad experience and we unblock Damian's PR.

dougbu commented:
> Since you obviously have strong opinions about how the analyzer should behave, I suggest you try it out.

Did some games w/ Microsoft.Web.Xdt.Extensions.csproj and couldn't introduce duplicates in that project's PublicAPI.Shipped.txt file. So we're left w/ deciding whether to move forward w/ Microsoft.CodeAnalysis.PublicApiAnalyzers and this PR or try another API governance approach.

To be clear, my concern about Microsoft.CodeAnalysis.PublicApiAnalyzers isn't exactly the self-supported point @jmarolf made, it's the lack of servicing strategy and murky versioning. It seems the package is not only self-supported but also ungoverned.

wtgodbe commented:
I also have some concerns about continuing to use the analyzer if it isn't going to be supported - @jmarolf @sharwell are there plans to support the roslyn API analyzer long-term?

I think we should take this PR once it's green, then talk about how we want to do API surface area verification going forward in our next DOI meeting.

dougbu commented:
Sorry, coming back to this after a bit. Other than rebasing on 'main', any changes in the first two commits @BrennanConroy❔

dougbu commented:
Oh, the "new but old" default constructors remain in PublicApi.Unshipped.txt files. As @wtgodbe suggested, those should all or mostly be n Shipped files.

dougbu commented on src/Security/Authentication/WsFederation/src/PublicAPI.Shipped.txt:
```diff
@@ -131,6 +77,8 @@ Microsoft.AspNetCore.Authentication.WsFederation.WsFederationHandler
 Microsoft.AspNetCore.Authentication.WsFederation.WsFederationHandler.Events.set -> void
 Microsoft.AspNetCore.Authentication.WsFederation.WsFederationHandler.WsFederationHandler(Microsoft.Extensions.Options.IOptionsMonitor<Microsoft.AspNetCore.Authentication.WsFederation.WsFederationOptions!>! options, Microsoft.Extensions.Logging.ILoggerFactory! logger, System.Text.Encodings.Web.UrlEncoder! encoder, Microsoft.AspNetCore.Authentication.ISystemClock! clock) -> void
+~Microsoft.AspNetCore.Authentication.WsFederation.WsFederationOptions.ConfigurationManager.get -> Microsoft.IdentityModel.Protocols.IConfigurationManager<Microsoft.IdentityModel.Protocols.WsFederation.WsFederationConfiguration!>!
+~Microsoft.AspNetCore.Authentication.WsFederation.WsFederationOptions.ConfigurationManager.set -> void
```
Where did these supposedly non-null-aware additions come from❔

dougbu commented on src/Http/Http.Features/src/PublicAPI.Shipped.txt:
```diff
@@ -1,6 +1,19 @@
 #nullable enable
-~Microsoft.AspNetCore.Http.Features.FeatureReference<> (forwarded, contained in Microsoft.Extensions.Features)
-~Microsoft.AspNetCore.Http.Features.FeatureReferences<> (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReference<T> (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReference<T>.Fetch(Microsoft.AspNetCore.Http.Features.IFeatureCollection! features) -> T? (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReference<T>.Update(Microsoft.AspNetCore.Http.Features.IFeatureCollection! features, T feature) -> T (forwarded, contained in Microsoft.Extensions.Features)
+static readonly Microsoft.AspNetCore.Http.Features.FeatureReference<T>.Default -> Microsoft.AspNetCore.Http.Features.FeatureReference<T> (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReference<T>.FeatureReference() -> void (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache> (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Cache -> TCache? (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Collection.get -> Microsoft.AspNetCore.Http.Features.IFeatureCollection! (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.FeatureReferences(Microsoft.AspNetCore.Http.Features.IFeatureCollection! collection) -> void (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Fetch<TFeature, TState>(ref TFeature? cached, TState state, System.Func<TState, TFeature?>! factory) -> TFeature? (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Fetch<TFeature>(ref TFeature? cached, System.Func<Microsoft.AspNetCore.Http.Features.IFeatureCollection!, TFeature?>! factory) -> TFeature? (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Initalize(Microsoft.AspNetCore.Http.Features.IFeatureCollection! collection) -> void (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Initalize(Microsoft.AspNetCore.Http.Features.IFeatureCollection! collection, int revision) -> void (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.FeatureReferences() -> void (forwarded, contained in Microsoft.Extensions.Features)
+Microsoft.AspNetCore.Http.Features.FeatureReferences<TCache>.Revision.get -> int (forwarded, contained in Microsoft.Extensions.Features)
```
I think you mentioned this earlier but to confirm: Forwarding a type no longer implicitly forwards its members❔ That's very messy if so because it's not possible to forward a subset of the members w/o something weird going on (maybe partial classes but I doubt it). Feels like a regression in the analyzer.

dougbu commented on src/Middleware/HeaderPropagation/src/PublicAPI.Unshipped.txt:
```diff
@@ -1,4 +1,2 @@
 #nullable enable
-*REMOVED*~Microsoft.AspNetCore.HeaderPropagation.HeaderPropagationMiddleware.HeaderPropagationMiddleware(Microsoft.AspNetCore.Http.RequestDelegate! next, Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.HeaderPropagation.HeaderPropagationOptions!>! options, Microsoft.AspNetCore.HeaderPropagation.HeaderPropagationValues! values) -> void
-Microsoft.AspNetCore.HeaderPropagation.HeaderPropagationMiddleware.HeaderPropagationMiddleware(Microsoft.AspNetCore.Http.RequestDelegate! next, Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.HeaderPropagation.HeaderPropagationOptions!>! options, Microsoft.AspNetCore.HeaderPropagation.HeaderPropagationValues! values) -> void
```
The third commit contains lots of obvious changes that just remove the leading tilde. Are removals like this due to duplicates between the Shipped and Unshipped files that weren't handled in earlier commits❔

dougbu reviewed (commented):
I _think_ I understand most of the third commit though

1. It's not clear to me why there remains so much in the PublicAPI.Unshipped.txt files. Did that much really change since 6.0.0❔
2. I don't know the code nor what's actually changed aside from nullability touch-ups that should now be handled exclusively in the PublicAPI.Shipped.txt files. Suggest we need area owners to look closely at the PublicAPI.Unshipped.txt files.

BrennanConroy commented:
> It's not clear to me why there remains so much in the PublicAPI.Unshipped.txt files. Did that much really change since 6.0.0❔

Yes? Besides the struct ctors, there were a bunch of new APIs.

> Forwarding a type no longer implicitly forwards its members❔

I don't know the exact behavior here, but the fix I found was to explicitly add any new API in the forwarded type to the old forward-from assembly as well as the actual location the type lives now.

> The third commit contains lots of obvious changes that just remove the leading tilde. Are removals like this due to duplicates between the Shipped and Unshipped files that weren't handled in earlier commits❔

I think most of the changes were due to APIs either being marked as nullable in 7.0 or having nullable changes (such as when Extensions was nullable attributed and affected a lot of our APIs). The new version of the analyzer does not like having the same API only differ by nullability it seems like.

BrennanConroy committed:
fixup

BrennanConroy commented:
Default ctors moved to the shipped file.

wtgodbe reviewed (approved):
Changes _seem_ reasonable. Will there be manual changes needed going forward? Might be worth letting people know that any active changes will need to be rebased/regenerated

dougbu reviewed (approved):
Ugh, sorry this took longer than I expected. Nothing much to comment on 😸

BrennanConroy committed:
fb

DamianEdwards commented:
This is looking close 🎉

BrennanConroy commented:
You'll have to merge @dougbu to override the Code Check task

dougbu commented:
I'll check in a few hours (have an appointment I need to leave for soon) and get this in before @JamesNK nullifies something else 😃
